### ISSUE TITLE
ObservableFixedSizedRingBuffer rare ArgumentOutOfRange exception

### ISSUE THREAD

**Issue by Nathan-Morley-JS** [Cysharp/ObservableCollections#73]
I have managed to produce a seemingly rare exception using the ObservableFixedSizedRingBuffer.
I do not understand why it occurs, but I have produced a minimal reproduction of the bug below.
It involves creating an ISynchronizedView with a filter over the ObservableFixedSizedRingBuffer, and then creating an INotifyCollectionChangedSynchronizedViewList from the view, which is then discarded without disposing.

```
using ObservableCollections;

namespace Sandbox;

internal class Program
{
    static void Main(string[] args)
    {
        var buffer = new ObservableFixedSizeRingBuffer<int>(5);

        var view = buffer.CreateView(value => value);
        view.AttachFilter(value => value % 2 == 1);

        {
            // INotifyCollectionChangedSynchronizedViewList created from ISynchronizedView with a filter.
            var collection = view.ToNotifyCollectionChanged(); 

            // Not disposed here.
        }

        buffer.AddFirst(1);
        buffer.AddFirst(1);
        buffer.AddFirst(2);
        buffer.AddFirst(3);
        buffer.AddFirst(5);
        buffer.AddFirst(8); // Argument out of range
        buffer.AddFirst(13);
    }
}
```

I understand disposal is the correct usage of the collection, however, the bug was rather vague and took a few hours to track down, the least that could happen is improve the ergonomics around the exception to help the user fix it on their end.

**Comment by neuecc** [comment]
Thank you, that's a bug where the index gets out of sync when filtered.
It was handled in ObservableList, but there was an implementation oversight.
I'll fix it.

**Comment by neuecc** [comment]
Oh, no, it wasn't working correctly even with ObservableList.
I'll have to fix it.
Thanks!

**Comment by neuecc** [comment]
I've released v3.0.5, include fixed it.
